### DISCUSSION TITLE
Overall approach to data import & export

### DISCUSSION THREAD

**Discussion by mikekidner** [gaphor/gaphor#2403] [General]
Getting data in and out of a model is obviously critical to its usefulness. There appears to be a couple of places where this is sort of implemented, e.g. XMI export, diagram export,  but I've not found much else yet. I have started to write some scripts to do things like:

-  pull out a table of requirements & coverage, 
-  export diagrams based on regex name matching,  
-  import requirements from CSV data etc. 
-  get a list of parts
- apply a stylesheet from css file

Is there an existing framework/module/roadmap I should base these on, to make them more coherent contributions?

**Comment by mikekidner** [comment]
The docs about [services](https://docs.gaphor.org/en/latest/gaphor-services.html) gives a couple of additional ideas on navigating through a model.
And the [diagram export CLI](https://github.com/gaphor/gaphor/issues/1847)

**Comment by amolenaar** [comment]
Hi @mikekidner,

I agree that interacting with the model (programmatically) is crucial. That's why we've separated all UI code cleanly from the non-ui code.

We do not have a clear framework/roadmap. The options are plenty, everything from documentation, system configuration, to testing. So far I've been using Jupyter notebooks (e.g. for the services example you mention).

It would be really cool if we can have some more examples (tutorials) on how you can use models and how they can integrate with the development process.

**Reply by mikekidner** [reply]
Is the Examples folder the best place to add some scripts? Then update docs to reference those?

**Reply by danyeaw** [reply]
@mikekidner yup, that sounds right.

**Comment by danyeaw** [comment]
I think it would be fantastic if we also updated our gaphor.org homepage to highlight being able to do Interactive Model Development with Code.

**Comment by mikekidner** [comment]
@danyeaw / @amolenaar / @marek-piirikivi  we've been working on some scripts to pull data from models, you can find the code at 
https://bitbucket.org/resonatesystems/gaphortools

The main script is SRC/gaphortools/extract_docs.py

It is pretty embryonic at this stage but would be interested to hear your opinion on how to align it better with gaphors' design, and more efficient ways to interact with the model.

Cheers, Mike.
